## Gather

In [1]:
import pandas as pd
patients = pd.read_csv('./auralin_dataset/patients.csv')
treatments = pd.read_csv('./auralin_dataset/treatments.csv')
adverse_reactions = pd.read_csv('./auralin_dataset/adverse_reactions.csv')

## Assess

In [2]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [3]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [4]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [9]:
# list all

# create a series object out of all the columns and list the duplicates

all_cols = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))

all_cols[all_cols.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [11]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [14]:
# sort the values of weight, look for extreme / incorrect values

patients.weight.sort_values()
# looks like we have weights

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [15]:
# Zaitseva has a weight of 48
patients[patients['surname']=='Zaitseva']


patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state  zip_code        country  \
210  Wooster    OH   44691.0  United States   

                                       contact   birthdate  weight  height  \
210  330-202-2145CamillaZaitseva@superrito.com  11/26/1938    48.8      63   

      bmi  
210  19.1

In [16]:
sum(treatments.auralin.isnull())

0

In [17]:
sum(treatments.novodra.isnull())

0

# Clean

In [18]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

## Missing Data

### Missing Data : Treatments table

### define
There should be 350 rows in the treatments table however there are only 280, the missing values are in treatments_cut.csv

In [19]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [21]:
treatment_cut = pd.read_csv('./auralin_dataset/treatments_cut.csv')
treatment_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

join the 2 datasets together, using concatenate
### code

In [22]:
treatments_clean_full = pd.concat([treatments_clean, treatment_cut])

### test

In [25]:
treatments_clean_full.count()

given_name      350
surname         350
auralin         350
novodra         350
hba1c_start     350
hba1c_end       350
hba1c_change    213
dtype: int64

### Missing Data : Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)

Missing data in the hba1c_change column, this is calculated from hba1c_start and hba1c_end

### Test

In [43]:
# 109 rows contain NaN 
treatments[treatments['hba1c_change'].isnull()]

given_name        surname    auralin    novodra  hba1c_start  \
0        veronika       jindrová  41u - 48u          -         7.63   
2        yukitaka       takenaka          -  39u - 36u         7.68   
8           saber         ménard          -  54u - 54u         8.08   
9            asia        woźniak  30u - 36u          -         7.76   
10         joseph            day  29u - 36u          -         7.70   
12        roxanne      andreyeva  29u - 38u          -         9.54   
14         simone   baumgaertner  27u - 37u          -         7.74   
15           enco         žibrik  55u - 68u          -         7.78   
16        camilla       zaitseva  28u - 37u          -         7.53   
18     addolorata       lombardi          -  49u - 46u         7.75   
19         khalid       johnsrud          -  54u - 54u         8.35   
22    brancaleone          russo  53u - 60u          -         8.61   
23       chiemela      tobeolisa          -  43u - 47u         7.59   
31            bao            she          -  48u - 45u         9.09   
33         cannan        cabrera          -  40u - 31u         7.94   
34      alexander      mathiesen  47u - 58u          -         7.96   
36            chỉ            lâm  45u - 48u          -         7.68   
37      władysław      wieczorek  24u - 37u          -         7.92   
45         zdeněk          synek          -  31u - 31u         7.67   
47      jad allah          gaber          -  38u - 41u         7.97   
48            zak          kelly          -  38u - 38u         7.66   
50        albinca        komavec          -  41u - 39u         7.89   
51     kristoffer      martinsen  29u - 37u          -         9.18   
52             mỹ          quynh  57u - 64u          -         7.61   
54           oles        zhdanov  54u - 67u          -         7.52   
58         gabryŝ    tomaszewski  29u - 37u          -         7.87   
59      leixandre         alanis  61u - 67u          -         7.74   
60   onyekachukwu         obinna  37u - 46u          -         7.58   
62          hanka          gegič          -  19u - 27u         7.67   
68        fatimah          kinfe          -  43u - 42u         7.88   
..            ...            ...        ...        ...          ...   
204       maximus         henzen  49u - 55u          -         7.79   
205         najla          touma          -  39u - 44u         7.66   
208       katrine          lynge          -  43u - 42u         7.60   
209          kári     hervinsson  37u - 43u          -         8.09   
210       vanessa       ferguson  39u - 50u          -         7.97   
211         nương             vũ  31u - 38u          -         7.68   
216        þórunn  tryggvadóttir          -  30u - 32u         7.98   
218        sirkka      piirainen          -  26u - 24u         8.90   
224        cintia         csorba  47u - 50u          -         8.49   
230          eric             ek          -  51u - 47u         7.92   
235         eemil          laine          -  30u - 29u         7.89   
238          anja         hueber          -  31u - 32u         8.82   
240         rocco       christie          -  42u - 40u         7.96   
241    elisabetta        lorenzo          -  28u - 24u         7.78   
242        fareeq           kalb          -  51u - 54u         7.97   
244         angel          grant  31u - 37u          -         7.93   
246         tomáš       navrátil  24u - 36u          -         7.84   
247       johanna         dreher  51u - 59u          -         7.77   
248       brianna        lansell  39u - 47u          -         9.48   
251         aksel    vestergaard          -  42u - 38u         9.62   
253        albino      schiavone  35u - 43u          -         7.56   
254          jose          combs          -  39u - 36u         7.89   
258        csilla       herczegh          -  43u - 46u         7.71   
260         david       beauvais          -  26u - 23u         7.87   
261      caroline         shule

### Define

recalc all the values again to fill in the NaN values

### Code

In [45]:
treatments_clean_full['hba1c_change'] = treatments_clean_full['hba1c_start'] - treatments_clean_full['hba1c_end']

In [46]:
treatments_clean_full

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0      veronika      jindrová  41u - 48u          -         7.63       7.20   
1        elliot    richardson          -  40u - 45u         7.56       7.09   
2      yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3          skye    gormanston  33u - 36u          -         7.97       7.62   
4        alissa        montez          -  33u - 29u         7.78       7.46   
5       jasmine         sykes          -  42u - 44u         7.56       7.18   
6        sophia        haugen  37u - 42u          -         7.65       7.27   
7         eddie        archer  31u - 38u          -         7.89       7.55   
8         saber        ménard          -  54u - 54u         8.08       7.70   
9          asia       woźniak  30u - 36u          -         7.76       7.37   
10       joseph           day  29u - 36u          -         7.70       7.19   
11    kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12      roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13       gregor          bole          -  47u - 45u         7.61       7.16   
14       simone  baumgaertner  27u - 37u          -         7.74       7.30   
15         enco        žibrik  55u - 68u          -         7.78       7.34   
16      camilla      zaitseva  28u - 37u          -         7.53       7.13   
17         gina          cain          -  36u - 36u         7.88       7.40   
18   addolorata      lombardi          -  49u - 46u         7.75       7.33   
19       khalid      johnsrud          -  54u - 54u         8.35       7.94   
20         mile        stanić          -  47u - 48u         7.66       7.24   
21        tekla       walczak  29u - 39u          -         7.61       7.29   
22  brancaleone         russo  53u - 60u          -         8.61       8.18   
23     chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24         isac          berg  31u - 41u          -         9.68       9.29   
25       benoît        bonami          -  44u - 43u         9.82       9.40   
26       suhaim         rahal          -  49u - 47u         7.94       7.50   
27       mizuki         iwata          -  45u - 46u         7.70       7.23   
28      clinton        miller  42u - 51u          -         7.79       7.40   
29       eugene       mironov  42u - 49u          -         7.81       7.48   
..          ...           ...        ...        ...          ...        ...   
40      eufemio       rosario          -  37u - 40u         7.54       7.26   
41     dalmacia        madrid          -  26u - 23u         7.67       7.21   
42        daimy         tromp          -  40u - 45u         9.41       8.94   
43   hans peter      radislav  42u - 53u          -         7.60       7.20   
44       jeremy       montagu          -  52u - 52u         7.68       7.36   
45    štefanija       novosel  37u - 42u          -         7.64       7.17   
46         anja       mueller  43u - 56u          -         7.58       7.29   
47       regolo         nucci  51u - 59u          -         7.53       7.02   
48      nebechi     ekechukwu          -  37u - 39u         7.78       7.39   
49      satsita     batukayev          -  42u - 42u         7.63       7.25   
50      timothy        cotton          -  26u - 25u         7.92       7.52   
51      bjørnar        nilsen          -  36u - 33u         7.99       7.70   
52       vivien        zsinkó  30u - 38u          -         8.35       7.92   
53        genet      fesahaye  25u - 35u          -         7.73       7.36   
54        borna     šlezinger          -  42u - 41u         7.55       7.18   
55         mary         adams          -  32u - 33u         7.65       7.26   
56       blanka      jurković  31u - 43u          -         7.77       7.33   
57       joseph        tucker  48u - 56u          -         7.67       7.30   
58  christopher      woodward          -  55u - 51u    

## Tidiness Data

### Missing Data : Contact column in patients table contains two variables: phone number and email

In [60]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

### Define

- split the contact column, it contains both 

In [90]:
patients_clean['email'] = patients_clean['contact'].str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])')
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

C:\tools\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  """Entry point for launching an IPython kernel.


In [85]:
patients_clean

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [91]:
patients_clean.drop('contact', axis=1)

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [92]:
# test emai;s by checking the email column doesn't start with a number or + (sort by values will reveal this)

patients_clean.email.sort_values()

404                 AaliyahRice@dayrep.com
11            Abdul-NurMummarIsa@rhyta.com
332                  AbelEfrem@fleckens.hu
258                AbelYonatan@teleworm.us
305      AddolorataLombardi@jourrapide.com
118           AdibMutazzGhanem@fleckens.hu
420                AdlanShishani@gustr.com
238          AkselHVestergaard@armyspy.com
61                    AlanMilne@dayrep.com
130            AlbertRWolfe@jourrapide.com
409          AlbikaIbragimov@superrito.com
196             AlbinaZetticci@teleworm.us
37                AlbincaKomavec@rhyta.com
295            AlbinoSchiavone@fleckens.hu
276                AlexCrawford@dayrep.com
70       AlexanderEMathiesen@superrito.com
496         AlexanderHueber@jourrapide.com
56            AlissaMontezFranco@gustr.com
172              AlvinAJackson@armyspy.com
437              AlwinSvensson@armyspy.com
358          AmalieJChristensen@einrot.com
143    AmandaCavalcantiRibeiro@fleckens.hu
47               AnaSousaCorreia@rhyta.com
442     Ana